drivers/kscan/kscan_matrix_poll.c

```diff
@@ -0,0 +1,323 @@
+			/* Index all they keys that changed for each row
+			 * in order to debounce each key in terms of it
+			 */
+			if (row_changed & BIT(r)) {
+				kb_poll.scan_cycle_idx[c][r] = kb_poll.scan_cycles_idx;
+			}
+		}
+
+		kb_poll.matrix_unstable_state[c] |= row_changed;
+		kb_poll.matrix_previous_state[c] = matrix_new_state[c];
+	}
+
+	for (int c = 0; c < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; c++) {
+		deb_col = kb_poll.matrix_unstable_state[c];
+
+		if (!deb_col) {
+			continue;
+		}
+
+		/* Debouncing for each row key occurs here */
+		for (int r = 0; r < CONFIG_KSCAN_MATRIX_MAX_ROWS; r++) {
+			uint8_t mask = BIT(r);
+			uint8_t row_bit = matrix_new_state[c] & mask;
+
+			/* Continue if we already debounce a key */
+			if (!(deb_col & mask)) {
+				continue;
+			}
+
+			/* Convert the clock cycle differences to usec */
+			uint32_t debt = CLOCK_32K_HW_CYCLES_TO_US(
+				cycles_now - kb_poll.scan_clk_cycle[kb_poll.scan_cycle_idx[c][r]]);
+
+			/* Does the key requires more time to be debounced? */
+			if (debt < (row_bit ? kb_poll.deb_time_press : kb_poll.deb_time_rel)) {
+				/* Need more time to debounce */
+				continue;
+			}
+
+			kb_poll.matrix_unstable_state[c] &= ~row_bit;
+
+			/* Check if there was a change in the stable state */
+			if ((kb_poll.matrix_stable_state[c] & mask) == row_bit) {
+				/* Key state did not change */
+				continue;
+			}
+
+			/* The current row has been debounced, therefore update
+			 * the stable state. Then, proceed to notify the
+			 * application about the keys pressed.
+			 */
+			kb_poll.matrix_stable_state[c] ^= mask;
+			kb_poll.config.callback(dev, r, c, row_bit ? true : false);
+		}
+	}
+
+	return key_pressed;
+}
+
+static bool poll_expired(uint32_t start_cycles, int64_t *timeout)
+{
+	uint32_t stop_cycles;
+	uint32_t cycles_spent;
+	uint32_t microsecs_spent;
+
+	stop_cycles = k_cycle_get_32();
+	cycles_spent = stop_cycles - start_cycles;
+	microsecs_spent = CLOCK_32K_HW_CYCLES_TO_US(cycles_spent);
+
+	/* Update the timeout value */
+	*timeout -= microsecs_spent;
+
+	return *timeout >= 0;
+}
+
+void polling_task(void *dummy1, void *dummy2, void *dummy3)
+{
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period;
+	int64_t local_poll_timeout = kb_poll.poll_timeout;
+
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (true) {
+		kscan_resume_matrix_detection(kscan_dev, true);
+		kscan_drive_matrix_column(kscan_dev, KEYBOARD_COLUMN_DRIVE_ALL);
+		k_sem_take(&poll_lock, K_FOREVER);
+		LOG_DBG("Start KB scan!!");
+
+		/* Start polling */
+		kscan_resume_matrix_detection(kscan_dev, false);
+		uint32_t start_poll_cycles = k_cycle_get_32();
+
+		while (atomic_get(kb_poll.config.enable_scan) == 1U) {
+			uint32_t start_period_cycles = k_cycle_get_32();
+
+			if (check_key_events(kscan_dev)) {
+				local_poll_timeout = kb_poll.poll_timeout;
+				start_poll_cycles = k_cycle_get_32();
+			} else if (!poll_expired(start_poll_cycles, &local_poll_timeout)) {
```

Comment:
Does it make sense to rework this polling timeout to use `sys_clock_timeout_end_calc()`?  That [routine](https://docs.zephyrproject.org/latest/reference/kernel/timing/clocks.html#subsystems-using-k-timeout-t) seems purposely written to support this use case.

It would be good to eliminate `CLOCK_32K_HW_CYCLES_TO_US` if you can.

Reply:
The latest patch used sys_clock_timeout_end_calc for the polling timeout.
Also, I replaced the CLOCK_32K_HW_CYCLES_TO_US macro with Zephyr kernel API "k_cyc_to_us_floor32"